src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Services/ExtensionAssemblyManager.cs

```diff
@@ -0,0 +1,202 @@
+            // but it requires changes to the intellicode vscode extension and here to access our IServiceBroker instance via MEF.
+            var starredCompletionsComponentDll = StarredCompletionAssemblyHelper.GetStarredCompletionAssemblyPath(serverConfiguration.StarredCompletionsPath);
+            Contract.ThrowIfFalse(TryGetOrCreateLoadContext(starredCompletionsComponentDll));
+        }
+
+        foreach (var assemblyFilePath in serverConfiguration.ExtensionAssemblyPaths)
+        {
+            if (TryGetOrCreateLoadContext(assemblyFilePath))
+            {
+                validExtensionAssemblies.Add(assemblyFilePath);
+            }
+        }
+
+        return new ExtensionAssemblyManager(directoryLoadContexts.ToImmutableDictionary(), assemblyFullNameToLoadContext.ToImmutableDictionary(), validExtensionAssemblies.ToImmutable());
+
+        bool TryGetOrCreateLoadContext(string assemblyFilePath)
+        {
+            // Verify that the path is something we can load.
+            // If it's not, log helpful error messages and no-op.  We do not want to take down the server if an extension fails to load.
+            if (!File.Exists(assemblyFilePath))
+            {
+                logger.LogError("Extension path {assemblyFilePath} does not exist", assemblyFilePath);
+                return false;
+            }
+
+            var directory = Path.GetDirectoryName(assemblyFilePath);
+            if (directory == null)
+            {
+                logger.LogError("Failed to get directory from {assemblyFilePath}", assemblyFilePath);
+                return false;
+            }
+
+            var assemblyFullName = AssemblyName.GetAssemblyName(assemblyFilePath).FullName;
+            if (assemblyFullNameToLoadContext.TryGetValue(assemblyFullName, out var existingContext))
+            {
+                // MEF relies on type full names (*without* their assembly names) to identify parts.
+                // If an extension assembly is added to the catalog twice, then we will almost certainly get duplicate MEF parts
+                // which breaks consumers who are only expecting one part.
+                //
+                // We validate this constraint here by checking for duplicate extension assembly full names.
+                logger.LogError(
+                    "{assemblyFilePath} with assembly name {assemblyName} conflicts with extension loaded from {existingContextName}",
+                    assemblyFilePath, assemblyFullName, existingContext.Name);
+                return false;
+            }
+
+            if (directoryLoadContexts.TryGetValue(directory, out var directoryContext))
+            {
+                assemblyFullNameToLoadContext.Add(assemblyFullName, directoryContext);
+                return true;
+            }
+
+            var loadContext = GetOrCreateDirectoryContext(directory, assemblyFilePath);
```

Comment:
```suggestion
            if (!directoryLoadContexts.TryGetValue(directory, out var directoryContext))
            {
                directoryContext = GetOrCreateDirectoryContext(directory, assemblyFilePath);
            }
            
```

Then there's not two paths both updating assemblyNameToLoadContext and returning true.